Combine TLS support and raw socket support into a single process. Thi…

pboothe opened the pull request:
Makes it so that a single server can support both TLS and unencrypted connections.  Alters the `test_queueing` unit test to queue up raw, websocket, and TLS websocket tests all in the same queue.  All tests pass.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt/18)
<!-- Reviewable:end -->

nkinkade commented on src/usage.c:
```diff
@@ -70,11 +70,11 @@ void srv_long_usage(char* info) {
   printf("                           server to open a socket to the client (MID, SFW),\n");
-  printf("                           and requires the --certificate and --private-key\n");
-  printf("                           options also be set.\n");
+  printf("                           and also require the --certificate and --private-key\n");
```
Tiny detail... the option is --private_key (underscore), so the comment is slightly off.

pboothe commented:
*[src/usage.c, line 76 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZWyRvbpHh2ArIcYHh-r1-76:-K4Zjf4LjJCSQjdK4L3M:-884325680) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/usage.c#L76)):*
Fixed in the next commit - thanks!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

pboothe commented:
*[src/web100srv_unit_tests.c, line 89 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZkRyUzR2Nr2nx78ql:-K4ZkRyUzR2Nr2nx78qm:2034156178) ([raw file](https://github.com/m-lab/ndt/blob/fbd69bc6dc6c99f2161361dcf794a9b6a1d42cb8/src/web100srv_unit_tests.c#L89)):*
Moved downwards to the bottom of the file, and it was changed to include a mix of raw, websocket, and secure websocket clients.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

pboothe commented:
*[src/web100srv_unit_tests.c, line 89 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZkRyUzR2Nr2nx78ql:-K4Zkn8fKSM4qCVasdDd:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/fbd69bc6dc6c99f2161361dcf794a9b6a1d42cb8/src/web100srv_unit_tests.c#L89)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

pboothe commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/web100srv_unit_tests.c, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZkK0fpz2-QZyJV-Jr:-K4ZkK0fpz2-QZyJV-Js:1752075637) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L15)):*
Refactored from run_ssl_test

---

*[src/web100srv_unit_tests.c, line 119 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZkuWjoyKogf8hMQLX:-K4ZkuWjoyKogf8hMQLY:306644915) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L119)):*
Refactored from test_node

---

*[src/web100srv_unit_tests.c, line 336 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZlHwO1yIIQMZcX5-c:-K4ZlHwO1yIIQMZcX5-d:-1926504229) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L336)):*
Tests are now ordered by runtime, with the tests that take the longest happening last.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

pboothe commented:
Added comments on reviewable to try and give context to changes in web100srv_unit_test.c  

Feel free to ignore if they are not helpful.

mtlynch commented:
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/usage.c, line 77 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4cl2cibw8rsmxfSVae:-K4cl2cibw8rsmxfSVaf:-1167750741) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/usage.c#L77)):*
The wording here is kind of confusing. It seems to say it can't do a test if the test requires the server to open a socket and the test requires the `--certificate` and `--private-key` options (which I don't think is what we mean). Also I hate the passive voice and think it is to be avoided wherever possible.

Original
>Use this port for TLS sockets to conduct the tests.
>Note: TLS tests can't do tests which require the server to open a socket to the client (MID, SFW), and also require the --certificate and --private-key options be set. If this is unset, TLS is not used.

Suggested

>Use this port to conduct the tests over TLS sockets.
>This option requires requires the --certificate and --private-key options be set. If this option is unset, NDT does not use TLS.
>Note: NDT cannot perform any tests over TLS if the test requires the server to open a socket to the client (MID, SFW).

---

*[src/web100srv.c, line 1553 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4cofq3jOrPdbLyL7R8:-K4cofq3jOrPdbLyL7R9:915093201) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1553)):*
Need to update the return part.

---

*[src/web100srv.c, line 1555 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4corcNEVvYZLaBq1wY:-K4corcNEVvYZLaBq1wZ:2009972930) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1555)):*
Need to document `fd_max`

---

*[src/web100srv.c, line 1891 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4crFW4O9Cbu89-Di9X:-K4crFW4O9Cbu89-Di9Y:-387106392) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1891)):*
It seems like this is ignored if `ssl_context` is `NULL`. Can we document this?

---

*[src/web100srv.c, line 1892 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4cqEpYK-IL8EzwDgcv:-K4cqEpYK-IL8EzwDgcw:1907698786) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1892)):*
Can we specify that these are non-TLS clients?

---

*[src/web100srv.c, line 1899 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dEH0ImTxi6dcNJOa9:-K4dEH0ImTxi6dcNJOaA:869275954) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1899)):*
Should be underscore separated, suggest `fd_max` for consistency with the function above.

---

*[src/web100srv.c, line 2323 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dFqDEAujRFQOHfg_j:-K4dFqDEAujRFQOHfg_k:1544039023) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L2323)):*
It's weird that one of these is a flag name and the other two are friendly names. Consider making them all the flag names.

---

*[src/web100srv.c, line 2398 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dG3kPY0JM2AJ9nJmg:-K4dG3kPY0JM2AJ9nJmh:-1699017084) ([raw file](https://github.com/m-lab/ndt/blob/fbd69bc6dc6c99f2161361dcf794a9b6a1d42cb8/src/web100srv.c#L2398)):*
What happened to this?

---

*[src/web100srv.c, line 2428 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dKiSp0tDEWdT6GLRX:-K4dKiSp0tDEWdT6GLRY:1519847178) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L2428)):*
Shouldn't the first parameter be `NULL` since it does not yet exist? This makes it explicit that we're not passing anything to `CreateListenSocket`.

---

*[src/web100srv.c, line 2429 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dLFuyGabj4VEgr4VY:-K4dLFuyGabj4VEgr4VZ:-296353682) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L2429)):*
Can we do the ternary outside the `if`? This is a lot of logic to cram into a single if condition.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Reviewed 1 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Review status: 2 of 3 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/web100srv_unit_tests.c, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZkK0fpz2-QZyJV-Jr:-K4dMdEzaqVXJV4Hpoa2:-760931637) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L15)):*
Whee refactoring!

---

*[src/web100srv_unit_tests.c, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dMh90aY1rlPoaueoP:-K4dMh90aY1rlPoaueoQ:779266736) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L15)):*
I don't understand what we mean by "passed-in specs". We're just passing in filenames.

---

*[src/web100srv_unit_tests.c, line 29 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dMQT9T101LNAyV_Nl:-K4dMQT9T101LNAyV_Nm:-1061515049) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L29)):*
Can we make these `snprintf` because now if the params  are long paths, we overflow the buffer.

---

*[src/web100srv_unit_tests.c, line 287 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dOrt3AIpvVp279EYR:-K4dOrt3AIpvVp279EYS:-1800821644) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L287)):*
We're repeating this code a lot (in the TLS and non-TLS tests). Suggest refactoring to a function like

```
choose_random_ports(int *port, int *tls_port) {
  srandom(time(NULL));
  port = (random() % 30000) + 1024;
  if (tls_port == NULL) return;
  do {
    tls_port = (random() % 30000) + 1024;
  } while (tls_port == port);
}
```

---

*[src/web100srv_unit_tests.c, line 336 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZlHwO1yIIQMZcX5-c:-K4dLlWj0zZo7ntOwmiT:457022496) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L336)):*
Can add a comment explaining this?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

pboothe commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/usage.c, line 77 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4cl2cibw8rsmxfSVae:-K4dMZQXHC04TFCm7iJw:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/usage.c#L77)):*
Done.

---

*[src/usage.c, line 110 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dZitjbuhmAgT8MbKz:-K4dZitjbuhmAgT8MbL-:42877363) ([raw file](https://github.com/m-lab/ndt/blob/03fe6e01779bb635070b2142a9b1460d47c33d17/src/usage.c#L110)):*
One cleanup - I added this option a long time ago but failed to document it in the usage.

---

*[src/web100srv.c, line 1553 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4cofq3jOrPdbLyL7R8:-K4dNgphKLCNyoqC1rnx:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1553)):*
Done.

---

*[src/web100srv.c, line 1555 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4corcNEVvYZLaBq1wY:-K4dNhOfGpHEnUh6jFGr:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1555)):*
Done.

---

*[src/web100srv.c, line 1891 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4crFW4O9Cbu89-Di9X:-K4dNsM6AQhaAW7bTfak:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1891)):*
Done.

---

*[src/web100srv.c, line 1892 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4cqEpYK-IL8EzwDgcv:-K4dO3CtmaC23IBHJYRT:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1892)):*
Done.

---

*[src/web100srv.c, line 1899 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dEH0ImTxi6dcNJOa9:-K4dOAhQ0LWl-J9w2OVo:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L1899)):*
Done.

---

*[src/web100srv.c, line 2323 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dFqDEAujRFQOHfg_j:-K4dOLnUSbK8rlJqxkUN:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L2323)):*
Done.

---

*[src/web100srv.c, line 2398 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dG3kPY0JM2AJ9nJmg:-K4dOMKsGjbfhiAetyma:-1317685753) ([raw file](https://github.com/m-lab/ndt/blob/fbd69bc6dc6c99f2161361dcf794a9b6a1d42cb8/src/web100srv.c#L2398)):*
It was dead code. Has been forever, I only noticed that it was dead just now when I went to create a parallel initialization for `tls_listenfd`.

---

*[src/web100srv.c, line 2428 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dKiSp0tDEWdT6GLRX:-K4dOgTbzIlhFa3sxQpr:844627677) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L2428)):*
Changed (both here and above)

---

*[src/web100srv.c, line 2429 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dLFuyGabj4VEgr4VY:-K4dPstLHKwHxXkb6DEQ:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv.c#L2429)):*
Done.

---

*[src/web100srv_unit_tests.c, line 15 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dMh90aY1rlPoaueoP:-K4dPyaP-iNWTyBgN5UR:-858773685) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L15)):*
The filenames we pass in have XXXXX in them for mkstemp to replace with other chars.

---

*[src/web100srv_unit_tests.c, line 29 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dMQT9T101LNAyV_Nl:-K4dQHisCXVsojpLi1B1:1215240901) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L29)):*
done, also bumped the buffer size

---

*[src/web100srv_unit_tests.c, line 287 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dOrt3AIpvVp279EYR:-K4dSQ_MxpvMlUfNzzOa:1817247884) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L287)):*
Done, but without making the memory point to a random location. :)

---

*[src/web100srv_unit_tests.c, line 336 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4ZlHwO1yIIQMZcX5-c:-K4dQeNvmbvwaRoqfWNk:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/7e89903bca3de3cd926ba093f16d86cf6a5b5ba5/src/web100srv_unit_tests.c#L336)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
LGTM

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/web100srv.c, line 2013 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dcR_Rnf6pv_7n9Z0m:-K4dcR_Rnf6pv_7n9Z0n:-481680163) ([raw file](https://github.com/m-lab/ndt/blob/03fe6e01779bb635070b2142a9b1460d47c33d17/src/web100srv.c#L2013)):*
Suggest something more descriptive like `socket_buffer_size`

---

*[src/web100srv_unit_tests.c, line 339 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/18#-K4dd5QM3ffs5IweqOur:-K4dd5QM3ffs5IweqOus:1983561661) ([raw file](https://github.com/m-lab/ndt/blob/03fe6e01779bb635070b2142a9b1460d47c33d17/src/web100srv_unit_tests.c#L339)):*
Suggest `in ascending order of expected runtime`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/m-lab/ndt/18#-:-K4ddDjHVYSL8b4DfPWZ:676834233)*
<!-- Sent from Reviewable.io -->
